controllers/openstackmachine_controller.go

```diff
@@ -506,6 +521,36 @@ func machineToInstanceSpec(openStackCluster *infrav1.OpenStackCluster, machine *
 	return &instanceSpec
 }
 
+var (
+	spaceDotUnderscore = regexp.MustCompile(`[ _\.]`)
+	notWordDotDash     = regexp.MustCompile(`[^\w.-]+`)
+)
+
+// sanitizeHostname mangles a server's name into a hostname. It is required to be functionally identical to the version in nova:
+//
+//	https://github.com/openstack/nova/blob/87d4807848bb9546c1fca972da2eb2eda13eb08d/nova/utils.py#L356-L389
+//
+// Do not change the behaviour of this function except to match the behaviour in nova.
```

Comment:
Is there any risk that we end up in a situation where nova has different behaviors for different versions?

Reply:
I would hope not. If it did that would be a potential upgrade issue for Nova itself: afaict it doesn't store the mangled value but calculates it when it's used, so changing the function would change the hostname for a running server, which probably isn't a good idea.

Perhaps @stephenfin has more insight.

Reply:
Unfortunately you already do have different behaviours. The code we're reimplemented here was changed relatively recently ([in Wallaby](https://github.com/openstack/nova/commit/9046f0fff4be424eda25401a3f9b8752964de775), to be precise). Prior to this, we did not replace the periods as part of the sanitisation. As such, you are going to see slightly different behaviour if this is deployed on an older cloud in situations where a user requests an instance name with periods in it e.g. `my-instance.example.com`.

Adding to the confusion, RHOSP includes a downstream-only workaround config option to opt-out of the new behaviour and disable the additional sanitization. IIRC this requires a support exception to use but it does exist. When this workaround config option is enabled, the cloud will behave like an older cloud. I don't recall what version of RHOSP this is present in.

Finally, Nova has the ability to request hostnames via the API [starting in microversion 2.90 (Xena)](https://docs.openstack.org/nova/latest/reference/api-microversion-history.html#maximum-in-xena-and-yoga). We presumably don't use this in CAPO right now, but if it ever is used then we should be careful to start using the `OS-EXT-SRV-ATTR:hostname` field returned in the API instead of this.

Reply:
Eurgh 😢 

I'm going to try to process this to see what it would mean for us.

Reply:
Ouch! So there would potentially be issues for consumers of this in cases where CAPO and Nova does not agree on the hostname then. :slightly_frowning_face: